### ISSUE TITLE
Terraform 1.9 incorrectly plans `update in-place` for an unchanged resource

### ISSUE THREAD

**Issue by ryancragun** [hashicorp/terraform#35424]
### Terraform Version

```shell
Terraform v1.9.1
on darwin_amd64
+ provider registry.terraform.io/hashicorp-forge/enos v0.5.2
```


### Terraform Configuration Files

```terraform
terraform {
  required_providers {
    enos = {
      source = "registry.terraform.io/hashicorp-forge/enos"
    }
  }
}

resource "enos_bundle_install" "vault" {
  destination = "/opt/bin/vault"

  artifactory = {
    username = "<your-username@hashicorp.com>"
    token    = "<your-artifactory-token>"
    url      = "https://artifactory.hashicorp.engineering/artifactory/hashicorp-packagespec-buildcache-local/cache-v1/vault-enterprise/7fb88d4d3d0a36ffc78a522d870492e5791bae1b0640232ce4c6d69cc22cf520/store/f45845666b4e552bfc8ca775834a3ef6fc097fe0-1a2809da73e5896b6f766b395ff6e1804f876c45.zip"
    sha256   = "d01a82111133908167a5a140604ab3ec8fd18601758376a5f8e9dd54c7703373"
  }

  transport = {
    ssh = {
      user             = "ubuntu"
      private_key_path = abspath("/path/to/ssh/key.pem")
      host             = "54.202.180.144"
    }
  }
}
```

### Debug Output

https://gist.github.com/ryancragun/f4ddef284e883cc56dcae0eee252fec8

### Expected Behavior

Terraform should plan with no updates

### Actual Behavior

Terraform plans an update even though no attributes have changed.

### Steps to Reproduce

This resource executes on a remote machine via SSH so you'll need to use something that has an IP address and SSH.

1. `terraform apply`
2. `terraform plan`

### Additional Context

A [test for the `enos` provider failed](https://github.com/hashicorp-forge/terraform-provider-enos/pull/18/checks#step:10:811) that has been working fine for years. I created the small repro above and tested with a released version of the provider and see the same behavior. The only thing that has changed is going from Terraform 1.8.5 to Terraform 1.9.1.

When testing the same module with Terraform 1.8.5 it does not see the resource as applyable, as one would expect.

The MsgPack bytes that are passed into provider request match the response exactly. (I snipped some of it out as it has credentials in it):
```
Error: The terraform-provider-enos plugin crashed!
                                                                                                                                                                                                                                   This is always indicative of a bug within the plugin. It would be immensely                                                                                                                                                        helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
                                                                                                                                                                                                                                                                                                                                                                                                                                                                      ╷                                                                                                                                                                                                                                  │ Error: exit status 1                                                                                                                                                                                                             │                                                                                                                                                                                                                                  │ Error: Request cancelled                                                                                                                                                                                                         │                                                                                                                                                                                                                                  │ The plugin6.(*GRPCProvider).PlanResourceChange request was cancelled.                                                                                                                                                            │
│ Stack trace from the terraform-provider-enos plugin:                                                                                                                                                                             │                                                                                                                                                                                                                                  │ panic: (*tfprotov6.PlanResourceChangeRequest)(0xc00069e4b0)({
│  TypeName: (string) (len=19) "enos_bundle_install",
│  PriorState: (*tfprotov6.DynamicValue)(0xc00028ae10)({
│   MsgPack: ([]uint8) (len=618 cap=640) {                                                                                                                                                                     13:19:21 [1396/5791]
│    00000000  86 ab 61 72 74 69 66 61  63 74 6f 72 79 84 a6 73  |..artifactory..s|
│    00000010  68 61 32 35 36 d9 40 64  30 31 61 38 32 31 31 31  |ha256.@d01a82111|
│    00000020  31 33 33 39 30 38 31 36  37 61 35 61 31 34 30 36  |133908167a5a1406|
│    00000030  30 34 61 62 33 65 63 38  66 64 31 38 36 30 31 37  |04ab3ec8fd186017|
│    00000040  35 38 33 37 36 61 35 66  38 65 39 64 64 35 34 63  |58376a5f8e9dd54c|
│    000000b0  74 74 70 73 3a 2f 2f 61  72 74 69 66 61 63 74 6f  |ttps://artifacto|
│    000000c0  72 79 2e 68 61 73 68 69  63 6f 72 70 2e 65 6e 67  |ry.hashicorp.eng|
│    000000d0  69 6e 65 65 72 69 6e 67  2f 61 72 74 69 66 61 63  |ineering/artifac|
│    000000e0  74 6f 72 79 2f 68 61 73  68 69 63 6f 72 70 2d 70  |tory/hashicorp-p|
│    000000f0  61 63 6b 61 67 65 73 70  65 63 2d 62 75 69 6c 64  |ackagespec-build|
│    00000100  63 61 63 68 65 2d 6c 6f  63 61 6c 2f 63 61 63 68  |cache-local/cach|
│    00000110  65 2d 76 31 2f 76 61 75  6c 74 2d 65 6e 74 65 72  |e-v1/vault-enter|
│    00000120  70 72 69 73 65 2f 37 66  62 38 38 64 34 64 33 64  |prise/7fb88d4d3d|
│    00000130  30 61 33 36 66 66 63 37  38 61 35 32 32 64 38 37  |0a36ffc78a522d87|
│    00000140  30 34 39 32 65 35 37 39  31 62 61 65 31 62 30 36  |0492e5791bae1b06|
│    00000150  34 30 32 33 32 63 65 34  63 36 64 36 39 63 63 32  |40232ce4c6d69cc2|
│    00000160  32 63 66 35 32 30 2f 73  74 6f 72 65 2f 66 34 35  |2cf520/store/f45|
│    00000170  38 34 35 36 36 36 62 34  65 35 35 32 62 66 63 38  |845666b4e552bfc8|
│    00000180  63 61 37 37 35 38 33 34  61 33 65 66 36 66 63 30  |ca775834a3ef6fc0|
│    00000190  39 37 66 65 30 2d 31 61  32 38 30 39 64 61 37 33  |97fe0-1a2809da73|
│    000001a0  65 35 38 39 36 62 36 66  37 36 36 62 33 39 35 66  |e5896b6f766b395f|
│    000001b0  66 36 65 31 38 30 34 66  38 37 36 63 34 35 2e 7a  |f6e1804f876c45.z|
│    000001c0  69 70 a8 75 73 65 72 6e  61 6d 65 b5 72 63 72 61  |ip.username.rcra|
│    000001d0  67 75 6e 40 68 61 73 68  69 63 6f 72 70 2e 63 6f  |gun@hashicorp.co|
│    000001e0  6d ab 64 65 73 74 69 6e  61 74 69 6f 6e ad 2f 6f  |m.destination./o|
│    000001f0  70 74 2f 72 65 61 6c 2f  61 72 74 a2 69 64 a6 73  |pt/real/art.id.s|
│    00000200  74 61 74 69 63 a4 70 61  74 68 c0 a7 72 65 6c 65  |tatic.path..rele|
│    00000210  61 73 65 c0 a9 74 72 61  6e 73 70 6f 72 74 92 c4  |ase..transport..|
│    00000220  2f 5b 22 6f 62 6a 65 63  74 22 2c 7b 22 73 73 68  |/["object",{"ssh|
│    00000230  22 3a 5b 22 6f 62 6a 65  63 74 22 2c 7b 22 68 6f  |":["object",{"ho|
│    00000240  73 74 22 3a 22 73 74 72  69 6e 67 22 7d 5d 7d 5d  |st":"string"}]}]|
│    00000250  81 a3 73 73 68 81 a4 68  6f 73 74 ae 35 34 2e 32  |..ssh..host.54.2|
│    00000260  30 32 2e 31 38 30 2e 31  34 34                    |02.180.144|
│   },
│   JSON: ([]uint8) <nil>                                                                                                                                                                                                          │  }),
 ProposedNewState: (*tfprotov6.DynamicValue)(0xc00028ae40)({
│   MsgPack: ([]uint8) (len=618 cap=640) {
│    00000000  86 ab 61 72 74 69 66 61  63 74 6f 72 79 84 a6 73  |..artifactory..s|
│    00000010  68 61 32 35 36 d9 40 64  30 31 61 38 32 31 31 31  |ha256.@d01a82111|
│    00000020  31 33 33 39 30 38 31 36  37 61 35 61 31 34 30 36  |133908167a5a1406|
│    00000030  30 34 61 62 33 65 63 38  66 64 31 38 36 30 31 37  |04ab3ec8fd186017|
│    00000040  35 38 33 37 36 61 35 66  38 65 39 64 64 35 34 63  |58376a5f8e9dd54c|
│    000000b0  74 74 70 73 3a 2f 2f 61  72 74 69 66 61 63 74 6f  |ttps://artifacto|
│    000000c0  72 79 2e 68 61 73 68 69  63 6f 72 70 2e 65 6e 67  |ry.hashicorp.eng|
│    000000d0  69 6e 65 65 72 69 6e 67  2f 61 72 74 69 66 61 63  |ineering/artifac|
│    000000e0  74 6f 72 79 2f 68 61 73  68 69 63 6f 72 70 2d 70  |tory/hashicorp-p|
│    000000f0  61 63 6b 61 67 65 73 70  65 63 2d 62 75 69 6c 64  |ackagespec-build|
│    00000100  63 61 63 68 65 2d 6c 6f  63 61 6c 2f 63 61 63 68  |cache-local/cach|
│    00000110  65 2d 76 31 2f 76 61 75  6c 74 2d 65 6e 74 65 72  |e-v1/vault-enter|
│    00000120  70 72 69 73 65 2f 37 66  62 38 38 64 34 64 33 64  |prise/7fb88d4d3d|
│    00000130  30 61 33 36 66 66 63 37  38 61 35 32 32 64 38 37  |0a36ffc78a522d87|
│    00000140  30 34 39 32 65 35 37 39  31 62 61 65 31 62 30 36  |0492e5791bae1b06|
│    00000150  34 30 32 33 32 63 65 34  63 36 64 36 39 63 63 32  |40232ce4c6d69cc2|
│    00000160  32 63 66 35 32 30 2f 73  74 6f 72 65 2f 66 34 35  |2cf520/store/f45|
│    00000170  38 34 35 36 36 36 62 34  65 35 35 32 62 66 63 38  |845666b4e552bfc8|
│    00000180  63 61 37 37 35 38 33 34  61 33 65 66 36 66 63 30  |ca775834a3ef6fc0|
│    00000190  39 37 66 65 30 2d 31 61  32 38 30 39 64 61 37 33  |97fe0-1a2809da73|
│    000001a0  65 35 38 39 36 62 36 66  37 36 36 62 33 39 35 66  |e5896b6f766b395f|
│    000001b0  66 36 65 31 38 30 34 66  38 37 36 63 34 35 2e 7a  |f6e1804f876c45.z|
│    000001c0  69 70 a8 75 73 65 72 6e  61 6d 65 b5 72 63 72 61  |ip.username.rcra|
│    000001d0  67 75 6e 40 68 61 73 68  69 63 6f 72 70 2e 63 6f  |gun@hashicorp.co|
│    000001e0  6d ab 64 65 73 74 69 6e  61 74 69 6f 6e ad 2f 6f  |m.destination./o|
│    000001f0  70 74 2f 72 65 61 6c 2f  61 72 74 a2 69 64 a6 73  |pt/real/art.id.s|
│    00000200  74 61 74 69 63 a4 70 61  74 68 c0 a7 72 65 6c 65  |tatic.path..rele|
│    00000210  61 73 65 c0 a9 74 72 61  6e 73 70 6f 72 74 92 c4  |ase..transport..|
│    00000220  2f 5b 22 6f 62 6a 65 63  74 22 2c 7b 22 73 73 68  |/["object",{"ssh|
│    00000230  22 3a 5b 22 6f 62 6a 65  63 74 22 2c 7b 22 68 6f  |":["object",{"ho|
│    00000240  73 74 22 3a 22 73 74 72  69 6e 67 22 7d 5d 7d 5d  |st":"string"}]}]|
│    00000250  81 a3 73 73 68 81 a4 68  6f 73 74 ae 35 34 2e 32  |..ssh..host.54.2|
│    00000260  30 32 2e 31 38 30 2e 31  34 34                    |02.180.144|
│   },
│   JSON: ([]uint8) <nil>
│  }),
```

### References

_No response_

**Comment by apparentlymart** [comment]
Hi @ryancragun! Thanks for reporting this, and for including the configuration and trace log.

Since this is about unexpected plan output, could you also share the plan that Terraform presented? Knowing exactly what Terraform was proposing to change should help narrow down what might be causing this. Thanks again!

**Comment by ryancragun** [comment]
Here ya go.

[plan.zip](https://github.com/user-attachments/files/16168936/plan.zip)

**Comment by apparentlymart** [comment]
Hi @ryancragun! Thanks for sharing that.

I had intended to ask for the human-readable plan output rather than literally the saved plan file, since saved plan files are not really designed to transfer alone between systems, but after a little finagling I was able to get a local copy of Terraform v1.9.1 to analyze this plan file enough to present it:

```
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # enos_bundle_install.vault will be updated in-place
  ~ resource "enos_bundle_install" "vault" {
        id          = "static"
        # (3 unchanged attributes hidden)
    }
```

As you noted, it does indeed seem like it's proposing a no-op change but yet showing it as an update anyway. Situations like this are often caused by some assumptions/heuristics in the UI code that didn't hold for this particular provider, and so one advantage of having started with your saved plan file is that I was also able to extract the lower-level JSON representation (what the UI works from) using `terraform show -json`:

1. Terraform apparently detected this object as having changed outside of Terraform, but the UI didn't show it because it didn't pass the heuristics we added in response to https://github.com/hashicorp/terraform/issues/28803:

	```json
	    {
	      "address": "enos_bundle_install.vault",
	      "mode": "managed",
	      "type": "enos_bundle_install",
	      "name": "vault",
	      "provider_name": "registry.terraform.io/hashicorp-forge/enos",
	      "change": {
		"actions": [
		  "update"
		],
		"before": {
		  "artifactory": {
		    "sha256": "d01a82111133908167a5a140604ab3ec8fd18601758376a5f8e9dd54c7703373",
		    "token": "DDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDD",
		    "url": "https://artifactory.hashicorp.engineering/artifactory/hashicorp-packagespec-buildcache-local/cache-v1/vault-enterprise/7fb88d4d3d0a36ffc78a522d870492e5791bae1b0640232ce4c6d69cc22cf520/store/f45845666b4e552bfc8ca775834a3ef6fc097fe0-1a2809da73e5896b6f766b395ff6e1804f876c45.zip",
		    "username": "rcragun@hashicorp.com"
		  },
		  "destination": "/opt/bin/vault",
		  "id": "static",
		  "path": null,
		  "release": null,
		  "transport": {
		    "ssh": {
		      "host": "54.202.180.144",
		      "private_key_path": "/Users/ryan/code/hashi/terraform-provider-enos/enos/support/enos-ci-ssh-key.pem",
		      "user": "ubuntu"
		    }
		  }
		},
		"after": {
		  "artifactory": {
		    "sha256": "d01a82111133908167a5a140604ab3ec8fd18601758376a5f8e9dd54c7703373",
		    "token": "DDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDD",
		    "url": "https://artifactory.hashicorp.engineering/artifactory/hashicorp-packagespec-buildcache-local/cache-v1/vault-enterprise/7fb88d4d3d0a36ffc78a522d870492e5791bae1b0640232ce4c6d69cc22cf520/store/f45845666b4e552bfc8ca775834a3ef6fc097fe0-1a2809da73e5896b6f766b395ff6e1804f876c45.zip",
		    "username": "rcragun@hashicorp.com"
		  },
		  "destination": "/opt/bin/vault",
		  "id": "static",
		  "path": null,
		  "release": null,
		  "transport": {
		    "ssh": {
		      "host": "54.202.180.144",
		      "private_key_path": "/Users/ryan/code/hashi/terraform-provider-enos/enos/support/enos-ci-ssh-key.pem",
		      "user": "ubuntu"
		    }
		  }
		},
		"after_unknown": {},
		"before_sensitive": {
		  "artifactory": true,
		  "transport": {
		    "ssh": {}
		  }
		},
		"after_sensitive": {
		  "artifactory": true,
		  "transport": {
		    "ssh": {}
		  }
		}
	      }
	    }
	```

	I don't notice any actual difference between the "before" and "after" values here, so it isn't clear to me why Terraform included this object in the "changes outside of Terraform" set. Perhaps this appearing here is a secondary symptom of the bug we're talking about.

2. Terraform Core also described a planned update for this resource instance:

	```json
	    {
	      "address": "enos_bundle_install.vault",
	      "mode": "managed",
	      "type": "enos_bundle_install",
	      "name": "vault",
	      "provider_name": "registry.terraform.io/hashicorp-forge/enos",
	      "change": {
		"actions": [
		  "update"
		],
		"before": {
		  "artifactory": {
		    "sha256": "d01a82111133908167a5a140604ab3ec8fd18601758376a5f8e9dd54c7703373",
		    "token": "DDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDD",
		    "url": "https://artifactory.hashicorp.engineering/artifactory/hashicorp-packagespec-buildcache-local/cache-v1/vault-enterprise/7fb88d4d3d0a36ffc78a522d870492e5791bae1b0640232ce4c6d69cc22cf520/store/f45845666b4e552bfc8ca775834a3ef6fc097fe0-1a2809da73e5896b6f766b395ff6e1804f876c45.zip",
		    "username": "rcragun@hashicorp.com"
		  },
		  "destination": "/opt/bin/vault",
		  "id": "static",
		  "path": null,
		  "release": null,
		  "transport": {
		    "ssh": {
		      "host": "54.202.180.144",
		      "private_key_path": "/Users/ryan/code/hashi/terraform-provider-enos/enos/support/enos-ci-ssh-key.pem",
		      "user": "ubuntu"
		    }
		  }
		},
		"after": {
		  "artifactory": {
		    "sha256": "d01a82111133908167a5a140604ab3ec8fd18601758376a5f8e9dd54c7703373",
		    "token": "DDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDDD",
		    "url": "https://artifactory.hashicorp.engineering/artifactory/hashicorp-packagespec-buildcache-local/cache-v1/vault-enterprise/7fb88d4d3d0a36ffc78a522d870492e5791bae1b0640232ce4c6d69cc22cf520/store/f45845666b4e552bfc8ca775834a3ef6fc097fe0-1a2809da73e5896b6f766b395ff6e1804f876c45.zip",
		    "username": "rcragun@hashicorp.com"
		  },
		  "destination": "/opt/bin/vault",
		  "id": "static",
		  "path": null,
		  "release": null,
		  "transport": {
		    "ssh": {
		      "host": "54.202.180.144",
		      "private_key_path": "/Users/ryan/code/hashi/terraform-provider-enos/enos/support/enos-ci-ssh-key.pem",
		      "user": "ubuntu"
		    }
		  }
		},
		"after_unknown": {},
		"before_sensitive": {
		  "artifactory": true,
		  "transport": {
		    "ssh": {}
		  }
		},
		"after_sensitive": {
		  "artifactory": true,
		  "transport": {
		    "ssh": {}
		  }
		}
	      }
	    }
	```

	Again I don't notice any clear difference between the before and after values, so it seems like we can rule out the idea that this is purely a UI-level concern: the underlying JSON data also seems to reflect a no-op action but presented as an "update". 

I notice you (wisely) redacted the "token" attribute's value before sharing this; can you confirm whether the original values of that _also_ match in before and after in your original? (You can generate similar output by running `terraform show -json PLANFILE` with your non-redacted saved plan artifact. The objects I showed above appear under `"resource_drift"` and `"resource_changes"` respectively.

I'm not sure yet what exactly to conclude from this information but the fact that the problem is also apparent in the underlying plan JSON -- rather than only in the human-oriented UI -- suggests that this is likely a Terraform Core bug rather than a UI-level bug.

Specifically, it seems that Terraform Core's logic has decided that the before and after values are non-equal, which would therefore cause the action type to be set to "update", even though there doesn't appear to be any difference between these values that we can observe in the JSON serialization. That could either mean that these values are different in some way that the JSON cannot distinguish -- e.g. both map and object types serialize as JSON objects -- or the equality check itself is somehow buggy. :thinking: 

I don't have any further ideas for now, but hopefully the context above is useful for someone else thinking about what might be going on here.

**Comment by apparentlymart** [comment]
The following is the logic for deciding what action to propose based on the prior and planned values in Terraform v1.9.1:

https://github.com/hashicorp/terraform/blob/6b1c106cfa05c954f27ca71a2a869a396a076cdd/internal/terraform/node_resource_abstract_instance.go#L2726-L2754

This suggests that the action should be "updated" only if the prior and planned values are not "equal" as defined by upstream `cty` -- the same definition that is used for the `==` operator in the Terraform language -- or if the planned object contains at least one unknown value somewhere in it which causes the equality test itself to return "unknown".

Again, I'm not really sure what to conclude from this yet but am sharing this context in the hope it will be useful for future investigation.

**Comment by jbardin** [comment]
OK, I just had an idea how to test and I think I've confirmed that this is related to the change in sensitivity handling that was present in v1.8. @ryancragun, was this resource state created prior to v1.8 and only now applying that update?

If that's the case there seems to be a minor regression, where Terraform could previously handle updating the sensitivity without showing an empty change in v1.8, but v1.9 now shows an empty update change while updating the state. Because there is no actual change in value, only that we store the sensitivity marks more precisely, is why there is no change shown for sensitivity in the human readable output.

**Comment by ryancragun** [comment]
@jbardin Nope, the resource was created by Terraform 1.9.1.

**Comment by jbardin** [comment]
I still don't know how this test was getting two different states, but I'm at least narrowing down the cause, sort of. The difference is that we have one state value where the schema marks were applied verbatim (which is how things were stored previously), and one state value where the marks were applied recursively within an object.

I have not yet been able to find where this is happening however. The schema still returns the same set of paths with `SensitivePaths` as it did for `ValueMarks`, which are both still eventually applied via `MarkWithPaths`. The end result though is that `marks.MarkPaths` is returning a different value than we used to get 😕 Maybe this rings a bell for @apparentlymart, since there does seem to be an inconsistency in mark handling somewhere.

**Comment by apparentlymart** [comment]
I'm afraid that nothing particularly interesting came to mind when I thought through what you just described, @jbardin. :confounded:

**Comment by jbardin** [comment]
Found the discrepancy! It originates with `go-cty` so I kind of surprised we've never hit it before (at least in a meaningful way).

The problem is within `MarkWithPaths`, but sort of inherent with anything that uses marks and `Transform` together. Terraform is assuming that running `Transform` with no changes will always return the same result, but the act of iterating over a marked value can change that value as marks are pushed down into nested structures.

For example if we start with
```
cty.ObjectVal(map[string]cty.Value{
  "nested":cty.ObjectVal(map[string]cty.Value{
    "attr":cty.StringVal("not directly marked"),
  }),
})
```

Running `MarkWithPaths` to mark the `nested` object returns:
```
cty.ObjectVal(map[string]cty.Value{
  "nested":cty.ObjectVal(map[string]cty.Value{
    "attr":cty.StringVal("not directly marked"),
  }).Mark("mark"),
})
```

But running `MarkWithPaths` a second time on the result returns
```
cty.ObjectVal(map[string]cty.Value{
  "nested":cty.ObjectVal(map[string]cty.Value{
    "attr":cty.StringVal("not directly marked").Mark("mark"),
  }).Mark("mark"),
})
```

While still semantically equal, these will return different sets of marked paths which are not being compared as equal. We can work around this in Terraform, but I'm not sure yet if we want to handle this more directly in `cty`.

**Comment by apparentlymart** [comment]
That is an interesting case, indeed. It's technically consistent with `cty`'s usual handling of marks in the small details -- the transform function receives a value that was accessed though an operation on another marked value and so the marks apply to the result -- but of course that then means that any values derived from the given value also end up marked, even though the value taken in isolation was not directly marked in the input.

The current interface for transformers doesn't give any other way to immediately know if an ancestor is marked, and so I think this comes down to a decision about whether the transform call for a nested value should receive that value directly (without any context about what it was nested inside) or whether it should obey the normal traversal rules.

With my `cty` maintainer hat on (instead of my Terraform developer hat), I notice that:
- A `cty.Transformer` implementation _could_ maintain its own state about which marks were present on ancestor objects if it needed to, by checking for them in `Enter` and building a stack of active marks that can be popped again during `Exit`, so there is still a way to get that information if you want it but it's considerably less convenient.
- The transform behavior is already actually inconsistent about how it handles this: it preserves ancestor marks when traversing into an attribute of an object, but it does not when traversing into an element of a collection. Therefore resolving this consistently one way or the other is definitely better than changing nothing.
- I accepted `MarkWithPaths` and its related API into `cty` primarily for Terraform's benefit, and so its behavior might as well reflect what Terraform needs, and it seems that _not_ propagating the marks downwards would be the most suitable behavior from Terraform's perspective.
- This seems like a pretty straightforward change, once I dig into the details: [the transforming internals use `val.GetAttr` today](https://github.com/zclconf/go-cty/blob/7b73cce468e8021d933cfb7990356837c6348146/cty/walk.go#L216), but could use `rawVal.GetAttr` instead to get the unmarked version -- which would be consistent with the other cases using `rawVal.ElementIterator` to iterate over collection or tuple elements -- and then the result should be what Terraform is expecting it to be. (There would still be [the call to mark the overall object at the end](https://github.com/zclconf/go-cty/blob/7b73cce468e8021d933cfb7990356837c6348146/cty/walk.go#L226), so the object-level markings would still be retained.

With all of that said then: I feel inclined to consider this inconsistency to be a bug in `cty.TransformWithTransformer` (and the related functions that share its implementation internals) and would be happy to accept a PR to fix it in a way similar to what I described above.

It is technically a slight breaking change for any caller that is relying on being able to get context from parent objects, but since it was working inconsistently anyway, and in practice I don't believe anything other than Terraform is currently making extensive use of this functionality, it seems better to me to clarify the behavior and then warn about the change in the release notes.
